src/codegen/available_expressions.rs

```diff
@@ -0,0 +1,866 @@
+            .borrow_mut()
+            .children
+            .insert(self.global_id_counter, Rc::clone(&new_node));
+
+        self.global_id_counter += 1;
+
+        self.global_id_counter - 1
+    }
+}
+
+impl AvailableExpressionSet {
+    /// Check if a commutative expression exists in the set
+    fn check_commutative(
+        &self,
+        exp: &Expression,
+        left: &Expression,
+        right: &Expression,
+    ) -> (bool, NodeId) {
+        let (valid_left, left_id) = self.find_expression(left);
+        let (valid_right, right_id) = self.find_expression(right);
+
+        if !valid_left || !valid_right {
+            return (false, 0);
+        }
+
+        let operator = get_operator_from_expression(exp);
+
+        if let Some(exp_id) = self.expr_map.get(&ExpressionType::BinaryOperation(
+            left_id, right_id, operator,
+        )) {
+            (true, *exp_id)
+        } else if let Some(exp_id) = self.expr_map.get(&ExpressionType::BinaryOperation(
+            right_id, left_id, operator,
+        )) {
+            (true, *exp_id)
+        } else {
+            (false, 0)
+        }
+    }
+
+    /// Add a commutative expression to the set if it is not there yet
+    fn process_commutative(
+        &mut self,
+        exp: &Expression,
+        left: &Expression,
+        right: &Expression,
+        ave: &mut AvailableExpression,
+    ) -> (bool, NodeId) {
+        let (valid_left, left_id) = self.gen_expression(left, ave);
+        let (valid_right, right_id) = self.gen_expression(right, ave);
+
+        if !valid_left || !valid_right {
+            return (false, 0);
+        }
+
+        let operator = get_operator_from_expression(exp);
+
+        if let Some(exp_id) = self.expr_map.get(&ExpressionType::BinaryOperation(
+            left_id, right_id, operator,
+        )) {
+            return (true, *exp_id);
+        } else if let Some(exp_id) = self.expr_map.get(&ExpressionType::BinaryOperation(
+            right_id, left_id, operator,
+        )) {
+            return (true, *exp_id);
+        }
+
+        (true, ave.add_expression_node(exp, self, left_id, right_id))
+    }
+
+    /// Get the hashmap key for a constant variable or a literal
+    fn constant_key(exp: &Expression) -> ConstantType {
+        match exp {
+            Expression::ConstantVariable(_, _, contract_no, var_no) => {
+                ConstantType::ConstantVariable(*contract_no, *var_no)
+            }
+
+            Expression::BytesLiteral(_, _, value) => ConstantType::Bytes(value.clone()),
+
+            Expression::BoolLiteral(_, value) => ConstantType::Bool(*value),
+
+            Expression::NumberLiteral(_, _, value) => ConstantType::Number(value.clone()),
+
+            _ => unreachable!("Not a constant"),
+        }
+    }
+
+    fn add_variable_or_arg(
+        &mut self,
+        exp: &Expression,
+        expr_type: &ExpressionType,
+        ave: &mut AvailableExpression,
+    ) -> (bool, NodeId) {
```

Comment:
The `bool` return is always true, not needed I guess

Reply:
Done